fdm-core/src/authorization.ts

```diff
@@ -0,0 +1,570 @@
+                        eq(authZSchema.role.resource, resource),
+                        eq(authZSchema.role.resource_id, resource_id),
+                        eq(authZSchema.role.principal_id, principal_id),
+                        eq(authZSchema.role.role, role),
+                        isNull(authZSchema.role.deleted),
+                    ),
+                )
+        })
+    } catch (err) {
+        throw handleError(err, "Exception for revoking role", {
+            resource: resource,
+            role: role,
+            resource_id: resource_id,
+            principal_id: principal_id,
+        })
+    }
+}
+
+export async function listResources(
+    fdm: FdmType,
+    resource: Resource,
+    action: Action,
+    principal_id: PrincipalId,
+): Promise<string[]> {
+    try {
+        // Get the roles for the action
+        const roles = getRolesForAction(action, resource)
+
+        // Convert principal_id to array
+        const principal_ids = Array.isArray(principal_id)
+            ? principal_id
+            : [principal_id]
+
+        // Query the resources available
+        const result = await fdm.transaction(async (tx: FdmType) => {
+            // Validate input
+            if (!resources.includes(resource)) {
+                throw new Error("Invalid resource")
+            }
+            if (!actions.includes(action)) {
+                throw new Error("Invalid action")
+            }
+
+            return await tx
+                .select({
+                    resource_id: authZSchema.role.resource_id,
+                })
+                .from(authZSchema.role)
+                .where(
+                    and(
+                        eq(authZSchema.role.resource, resource),
+                        inArray(authZSchema.role.principal_id, principal_ids),
+                        inArray(authZSchema.role.role, roles),
+                        isNull(authZSchema.role.deleted),
+                    ),
+                )
+        })
+        return result.map(
+            (resource: { resource_id: string }) => resource.resource_id,
+        )
+    } catch (err) {
+        throw handleError(err, "Exception for listing resources", {
+            resource: resource,
+            action: action,
+            principal_id: principal_id,
+        })
+    }
+}
+
+function getRolesForAction(action: Action, resource: Resource): Role[] {
+    const roles = permissions.filter((permission) => {
+        return (
+            permission.resource === resource &&
+            permission.action.includes(action)
+        )
+    })
+
+    const rolesFlat = roles.flatMap((role) => {
+        return role.role
+    })
+
+    return rolesFlat
+}
+
+async function getResourceChain(
+    fdm: FdmType,
+    resource: Resource,
+    resource_id: ResourceId,
+): Promise<ResourceChain> {
+    try {
+        const chainOrder = [
+            "farm",
+            "field",
+            "cultivation",
+            "harvesting",
```

Comment:
this order it not exactly followed in the ifelse lines below. by purpose?

Reply:
At the end it sorted by this order, so it does not have an impact on the performance, but maybe for code organization it is better to follow the same order